### ISSUE TITLE
Feature imp with xgboost: "Feature names stored in object and newdata are different!"

### ISSUE THREAD

**Issue by pat-s** [christophM/iml#46]
See https://stackoverflow.com/questions/51980808/r-plotting-importance-feature-using-featureimpnew/52056960#52056960 for a reprex.

Probably caused by internal disordering of the colnames due to shuffling.

Its a known issue that `xgboost` has problems with this: https://github.com/dmlc/xgboost/issues/1809

**Comment by christophM** [comment]
I could not replicate the issue with the code in the post. 
But I had the same problem when using `ranger::ranger`

```r
  library(mlr)
  library(ranger)
  data("iris")
  tsk = mlr::makeClassifTask(data = iris, target = "Species")
  lrn = mlr::makeLearner("classif.ranger",predict.type = "prob")
  mod = mlr:::train(lrn, tsk)
  X = iris[which(names(iris) != "Species")]
  predictor = Predictor$new(mod, data = X, y = iris$Species)
  imp = FeatureImp$new(predictor, loss = "ce")
```

@pat-s  could you test if it works for you now with xgboost?
Btw. reason was I sent the original data with some additional columns to the prediction function.

**Comment by gbouzill** [comment]
Just an additional note since I had the same issue.

The error occured because I had some additional preprocessing steps after the creation of the task, i.e:

```
tsk <- mlr::makeClassifTask(data = df, target = "y")
tsk <- createDummyFeatures(tsk, method = "1-of-n")
lrn <-  mlr::makeLearner("classif.ranger",predict.type = "prob")
mod <- mlr:::train(lrn, tsk)
```

The solution is to retrieve the data from the task with : 
`tsk$env$data`

**Comment by Aypak** [comment]
You need to ensure that the order of the columns is the same both when you train and predict. Repeat the pre-processing steps in the same order

**Comment by hubgitadi** [comment]
Two things I experienced with latest xgboost Version 1.1.1.1
- 1. order of columns should be the same in xgboost training and test data  
- 2. models saved using older xgboost versions with 'save'/ 'saveRDS' are not compatible with new versions, you need to use xgb.save and xgb.load for saving and retrieval respectively
